Content.Server/GameTicking/Rules/TraitorRuleSystem.cs

```diff
@@ -104,44 +116,66 @@ private void MakeCodewords()
         }
     }
 
-    private void OnPlayersSpawned(RulePlayerJobsAssignedEvent ev)
+    private void DoTraitorStart()
     {
-        if (!RuleAdded)
+        if (!_startCandidates.Any())
+        {
+            _sawmill.Error("Tried to start Traitor mode without any candidates.");
             return;
+        }
 
-        var numTraitors = MathHelper.Clamp(ev.Players.Length / _playersPerTraitor, 1, _maxTraitors);
+        var numTraitors = MathHelper.Clamp(_startCandidates.Count / _playersPerTraitor, 1, _maxTraitors);
         var codewordCount = _cfg.GetCVar(CCVars.TraitorCodewordCount);
 
-        var traitorPool = FindPotentialTraitors(ev);
+        var traitorPool = FindPotentialTraitors(_startCandidates);
         var selectedTraitors = PickTraitors(numTraitors, traitorPool);
 
         foreach (var traitor in selectedTraitors)
             MakeTraitor(traitor);
+
+        AnnouncementMade = true;
+    }
+
+    private void OnPlayersSpawned(RulePlayerJobsAssignedEvent ev)
+    {
+        if (!RuleAdded)
+            return;
+
+        foreach (var player in ev.Players)
+        {
+            if (!ev.Profiles.ContainsKey(player.UserId))
+                continue;
+
+            _startCandidates[player] = ev.Profiles[player.UserId];
+        }
+
+        var delay = TimeSpan.FromSeconds(
+            _cfg.GetCVar(CCVars.TraitorStartDelay) +
+            _random.NextFloat(0f, _cfg.GetCVar(CCVars.TraitorStartDelayVariance)));
+
+        _cancelStart = new CancellationTokenSource();
+        Timer.Spawn(delay, () => DoTraitorStart(), _cancelStart.Token);
```

Comment:
Something I haven't seen before. Is this similar to how events handle the startAfter field? That's what I would model it after.

Reply:
Looks like StationEvents just use an accumulator based on Update for each system.

Reply:
It shouldn't be too hard to hook into Update then use a comparison of the time instead of a spawned timer. I can make that change in a bit. I noticed that calling the emergency shuttle used Timer.Spawn so I figured that might be okay for here too.